### PR TITLE
[CALCITE-4199] Add nullability annotations to the methods and fields, ensure consistency with checkerframework

### PR DESCRIPTION
**Pull request by vlsi**
https://issues.apache.org/jira/browse/CALCITE-4199

CheckerFramework needs an annotated JDK (which might be slow to download), so it is not activated by default

Third-party issues:
* https://github.com/gradle/gradle/issues/14332: Order classes by recent failures and modification date when feeding them to javac compiler
* https://github.com/policeman-tools/forbidden-apis/issues/173: Ability to skip class file parse errors
* https://github.com/typetools/checker-framework/issues/3616: Infer annotations for receiver of instanceof checks
* https://github.com/typetools/checker-framework/issues/3614: False alarm when using "return null" and type is known to support nulls
* https://github.com/typetools/checker-framework/issues/3612: Crash in AbstractAtmComboVisitor visitArray_Typevar .. defaultAction

Corner cases:
* https://github.com/typetools/checker-framework/issues/399 checker framework does not understand `if (queue.isEmpty()) { ... = queue.poll(); }`, and it treats `poll()` as always nullable.
* Arrays vs annotations: https://twitter.com/tagir_valeev/status/1301443075080949760, https://twitter.com/tagir_valeev/status/1302173719398436864

### CODE REVIEW

**Comment by michaelmior** [comment]
Looks like this is currently failing but I assume that's expected? (Presumably because of missing annotations.)

**Comment by vlsi** [comment]
You are right, there's not enough annotations.
The idea was that once all the annotations are there, we could enforce the check for all the commits.

Feel free to pick it up.
I don't have the bandwidth in the nearest future :-/

**Comment by michaelmior** [comment]
@vlsi Unfortunately I'm in the same situation bandwidth-wise but I'll keep that in mind.

**Comment by zinking** [comment]
> You are right, there's not enough annotations.
> The idea was that once all the annotations are there, we could enforce the check for all the commits.
> 
> Feel free to pick it up.
> I don't have the bandwidth in the nearest future :-/

I remember you do these things using script last time I saw ^_^

**Comment by vlsi** [comment]
Just in case, I've added `nullability` annotations to `pgjdbc`.
Here's what it becomes: https://github.com/pgjdbc/pgjdbc/pull/1814

An interesting note is that I had to add several annotations for `java.sql` types (e.g. ResultSet, CallableStatement, Preparedstatement, ResultSet, etc)

**Comment by michaelmior** [comment]
@vlsi I was curious to take a look at this and I started to tackle some more necessary annotations on my own `checkerframework` branch. So far I see there are no `@NonNull` annotations in the codebase. Has this simply not been addressed yet or is there a reason for avoiding these?

**Comment by vlsi** [comment]
> So far I see there are no @NonNull annotations in the codebase

The suggested behavior is `non-null` by default, so I have not added it yet.

It should be enough to add `DefaultQualifier` to the topmost package as follows: https://github.com/pgjdbc/pgjdbc/blob/aa190cce2c27ac5f5d4fa33f5d156f8bf04cbe0c/pgjdbc/src/main/java/org/postgresql/package-info.java#L6-L13

**Comment by vlsi** [comment]
Just in case, `linq4j:compileJava` says there are only 46 errors left :)

**Comment by vlsi** [comment]
Ok, now `linq4j` passes the verification.

**Comment by uschindler** [comment]
Hi,
sorry for not producing a useful error message with forbidenapis. The problem here is that forbiddenapis catches a bug (indirectly) that is making Oracle's Javac produce bad class files.

After applying the forbiddenapis pull request https://github.com/policeman-tools/forbidden-apis/pull/174 I was able to find the problem:

- https://github.com/apache/calcite/blob/29bf6aff96325106756a989297e907daa9e2438d/core/src/main/java/org/apache/calcite/runtime/ResultSetEnumerable.java#L100
- https://github.com/apache/calcite/blob/29bf6aff96325106756a989297e907daa9e2438d/core/src/main/java/org/apache/calcite/runtime/ResultSetEnumerable.java#L439

In both cases javac creates buggy class files which may break downstream consumers of ResultSetEnumerable.

My recommendation is to remove the `@Nullable` here.

If you want to just work around the issue, you can exclude the file using "exclude" in the forbiddenapis configuration:

```
            configureEach<CheckForbiddenApis> {
                excludeJavaCcGenerated()
                exclude(
                    "**/org/apache/calcite/adapter/os/Processes${'$'}ProcessFactory.class",
                    "**/org/apache/calcite/adapter/os/OsAdapterTest.class",
                    "**/org/apache/calcite/runtime/Resources${'$'}Inst.class",
                    "**/org/apache/calcite/test/concurrent/ConcurrentTestCommandScript.class",
                    "**/org/apache/calcite/test/concurrent/ConcurrentTestCommandScript${'$'}ShellCommand.class",
                    "**/org/apache/calcite/util/Unsafe.class",
                    "**/org/apache/calcite/runtime/ResultSetEnumerable.class"
                )
            }
```

**Comment by vlsi** [comment]
@uschindler , thank you so much for the explanation. No need to sorry, and I need to say many thanks for the maintenance of forbidden-apis.

As you see, I am still in progress of annotating the code to verify nullness (there are 3000 errors to be fixed), and then I would have the full set of ill classes, and I could either refactor them or do something.

I do not think it is worth excluding ResultSetEnumerable from the verification. The reason for other classes are excluded is they have known forbidden calls which can't be avoided (e.g. the code is generated). By the way, if we could use "suppresswarnings" to suppress a single violation, we would go with that rather than excluding the full file.

**Comment by uschindler** [comment]
Hi,

If your're interested here is a patch to fix the issue. The good thing: it also rmeove the stupid cast, because Javac can't guess the lambda's types:

[javacbug.patch.txt](https://github.com/apache/calcite/files/5162279/javacbug.patch.txt)

If this also fixes checkerframework, go for it! Actually this seems another bug, because moving the type outside of the try...catch (see my code, does not help), but moving to a method helps.

To me this looks like a problem with javac and type parameters in combination with lambdas inside other lambdas. :-)

**Comment by uschindler** [comment]
Sorry the patch file still contains my quick hack to let forbiddenapis run from a local JAR file.

File: core/src/main/java/org/apache/calcite/runtime/ResultSetEnumerable.java [hunk 1]
```diff
@@ -95,29 +95,32 @@
           };
         } else {
           //noinspection unchecked
```
Comment (uschindler) [inline core/src/main/java/org/apache/calcite/runtime/ResultSetEnumerable.java:97]: is this comment not also obsolete? In my patch I removed it; reason: unchecked cast is gone

File: core/src/main/java/org/apache/calcite/runtime/ResultSetEnumerable.java [hunk 2]
```diff
@@ -434,20 +437,23 @@ public void close() {
         };
       }
       //noinspection unchecked
```
Comment (uschindler) [inline core/src/main/java/org/apache/calcite/runtime/ResultSetEnumerable.java:439]: same here

File: core/src/main/java/org/apache/calcite/DataContext.java [hunk 3]
```diff
@@ -42,17 +44,17 @@
   /**
    * Returns a sub-schema with a given name, or null.
    */
-  SchemaPlus getRootSchema();
+  @Nullable SchemaPlus getRootSchema();
 
   /**
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/DataContext.java:49]: Can we use `javax.annotation.Nullable` instead.

Comment (vlsi) [reply core/src/main/java/org/apache/calcite/DataContext.java:49]: `javax.annotation.Nullable` does not support cases like `List<javax.annotation.Nullable String>`

**Comment by vlsi** [comment]
As of now, 480 errors left in `:core`

**Comment by zinking** [comment]
@vlsi  not sure why this is closed. I think this will be useful and seems most agrees with this (JIRA). 
I'd suggest break these things into small iterations and put them in gradually.

**Comment by vlsi** [comment]
PR moved to https://github.com/apache/calcite/pull/2268

**Pull request closed without merging**